File: backend/routes_python/files.py
from fastapi import APIRouter, HTTPException
from fastapi.responses import FileResponse, Response
from pathlib import Path
import os
import mimetypes

router = APIRouter(prefix="/files", tags=["files"])

# Configure uploads directory
UPLOADS_DIR = Path(__file__).parent.parent / "uploads"
UPLOADS_DIR.mkdir(exist_ok=True)

@router.get("/{filename}")
async def serve_file(filename: str):
    """Serve uploaded files"""
    try:
        file_path = UPLOADS_DIR / filename
        
        # Check if file exists
        if not file_path.exists():
            raise HTTPException(status_code=404, detail="File not found")
        
        # Get MIME type
        mime_type, _ = mimetypes.guess_type(str(file_path))
        if not mime_type:
            mime_type = "application/octet-stream"
        
        # Return file with appropriate headers
        return FileResponse(
            path=file_path,
            media_type=mime_type,
            headers={
                "Cache-Control": "public, max-age=86400",  # Cache for 24 hours
                "ETag": f'"{filename}"'
            }
        )
        
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Internal server error: {str(e)}")

@router.delete("/{filename}")
async def delete_file(filename: str):
    """Delete uploaded file"""
    try:
        file_path = UPLOADS_DIR / filename
        
        if file_path.exists():
            os.remove(file_path)
            return {
                "success": True,
                "message": "File deleted successfully"
            }
        else:
            # File doesn't exist, consider it successful
            return {
                "success": True,
                "message": "File already deleted or not found"
            }
        
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Internal server error: {str(e)}")

@router.get("/admin/stats")
async def get_storage_stats():
    """Get storage statistics"""
    try:
        total_files = 0
        total_size = 0
        
        if UPLOADS_DIR.exists():
            for file_path in UPLOADS_DIR.iterdir():
                if file_path.is_file():
                    total_files += 1
                    total_size += file_path.stat().st_size
        
        return {
            "success": True,
            "storage": {
                "fileCount": total_files,
                "totalSize": total_size,
                "totalSizeMB": round(total_size / 1024 / 1024, 2),
                "uploadsDirectory": str(UPLOADS_DIR)
            }
        }
        
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Internal server error: {str(e)}")

@router.post("/admin/cleanup")
async def cleanup_old_files(days: int = 7):
    """Cleanup old files"""
    try:
        import time
        from datetime import datetime, timedelta
        
        cutoff_time = time.time() - (days * 24 * 60 * 60)
        deleted_count = 0
        
        if UPLOADS_DIR.exists():
            for file_path in UPLOADS_DIR.iterdir():
                if file_path.is_file() and file_path.stat().st_mtime < cutoff_time:
                    os.remove(file_path)
                    deleted_count += 1
        
        return {
            "success": True,
            "message": f"Cleanup completed for files older than {days} days",
            "deletedFiles": deleted_count
        }
        
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Internal server error: {str(e)}")

File: backend/routes/files.js
const express = require('express');
const fileService = require('../services/fileService');

const router = express.Router();

/**
 * Serve uploaded files
 * GET /api/files/:filename
 */
router.get('/:filename', async (req, res) => {
  try {
    const { filename } = req.params;

    // Get file from file service
    const file = await fileService.getFile(filename);

    // Set appropriate headers
    res.set({
      'Content-Type': file.mimeType,
      'Content-Length': file.size,
      'Cache-Control': 'public, max-age=86400', // Cache for 24 hours
      'ETag': `"${filename}"`
    });

    // Send file buffer
    res.send(file.buffer);

  } catch (error) {
    console.error('File serve error:', error);
    
    if (error.message === 'File not found') {
      return res.status(404).json({ 
        success: false, 
        error: 'File not found' 
      });
    }
    
    res.status(500).json({ 
      success: false, 
      error: 'Internal server error',
      message: error.message 
    });
  }
});

/**
 * Delete uploaded file
 * DELETE /api/files/:filename
 */
router.delete('/:filename', async (req, res) => {
  try {
    const { filename } = req.params;
    const { sessionId } = req.body;

    // In a real implementation, you'd verify the user has permission to delete
    // For now, we'll allow deletion with session ID verification
    
    const success = await fileService.deleteFile(filename);

    if (success) {
      res.json({
        success: true,
        message: 'File deleted successfully'
      });
    } else {
      res.status(500).json({
        success: false,
        error: 'Failed to delete file'
      });
    }

  } catch (error) {
    console.error('File delete error:', error);
    res.status(500).json({ 
      success: false, 
      error: 'Internal server error',
      message: error.message 
    });
  }
});

/**
 * Get storage statistics (admin endpoint)
 * GET /api/files/admin/stats
 */
router.get('/admin/stats', async (req, res) => {
  try {
    const stats = await fileService.getStorageStats();
    
    res.json({
      success: true,
      storage: stats
    });

  } catch (error) {
    console.error('Storage stats error:', error);
    res.status(500).json({ 
      success: false, 
      error: 'Internal server error',
      message: error.message 
    });
  }
});

/**
 * Cleanup old files (admin endpoint)
 * POST /api/files/admin/cleanup
 */
router.post('/admin/cleanup', async (req, res) => {
  try {
    const { days = 7 } = req.body;
    
    await fileService.cleanupOldFiles(days);
    
    res.json({
      success: true,
      message: `Cleanup completed for files older than ${days} days`
    });

  } catch (error) {
    console.error('Cleanup error:', error);
    res.status(500).json({ 
      success: false, 
      error: 'Internal server error',
      message: error.message 
    });
  }
});

module.exports = router;